dali/dfuplus/dfuplus.cpp

```diff
@@ -1503,112 +1524,135 @@ int CDfuPlusHelper::listhistory()
             {
-                // Get header from the first record
-                if (historyIter->first())
-                {
-                    Owned<IAttributeIterator> attribIter = historyIter->query().getAttributes(sorted);
-                    ForEach(*attribIter)
-                    {
-                        if (!first)
-                            csvDump.append(",");
-                        else
-                            first = false;
-
-                        const char *propName = attribIter->queryName();
-                        propName++;
-                        csvDump.append(propName);
-                    }
-                    csvDump.append("\n");
-                }
+                // header
+                csvDump.append("ip,name,operation,owner,path,timestamp,workunit\n");
             }
 
-            ForEach(*historyIter)
+            for(unsigned i = 0; i < arrHistory.length(); i++)
             {
-                first = true;
-                Owned<IAttributeIterator> attribIter = historyIter->query().getAttributes(sorted);
-                ForEach(*attribIter)
+                IConstHistory* historyRecord = &arrHistory.item(i);
+                if(historyRecord != nullptr)
                 {
-                    if (!first)
-                        csvDump.append(",");
-                    else
-                        first = false;
-
-                    const char *propVal =  attribIter->queryValue();
-                    csvDump.append(propVal);
+                    csvDump.appendf("%s,", historyRecord->getIP());
+                    csvDump.appendf("%s,", historyRecord->getName());
+                    csvDump.appendf("%s,", historyRecord->getOperation());
+                    csvDump.appendf("%s,", historyRecord->getOwner());
+                    csvDump.appendf("%s,", historyRecord->getPath());
+                    csvDump.appendf("%s,", historyRecord->getTimestamp());
+                    csvDump.appendf("%s\n",  historyRecord->getWorkunit());
                 }
-                csvDump.append("\n");
             }
             info("%s", csvDump.str());
         }
         break;
 
-    case json:
+        case json:
         {
-            StringBuffer out;
-            toJSON(history, out, 3);
-            info("\n%s\n",out.trim().str());
+            StringBuffer jsonDump;
+            const char * level1 = "    ";        // 4 space 'tab'
+            const char * level2 = "        ";
+            const char * level3 = "            ";
+            //                     123456789012
+
+            unsigned index = 1;
+
+            jsonDump.append("{\n");       // Level 0: Start History object
+
+            jsonDump.appendf("%s\"%s\": [", level1, "History");   // Begin History object's array of historical records
+
+            for(unsigned i = 0; i < arrHistory.length(); i++)
+            {
+                if (index++ != 1)
+                    jsonDump.append(",");
```

Comment:
This is fine, but I just want to mention if you use `delimitJSON(jsonDump);` you don't need the if.  Just looks slightly simpler.